expression.k

```diff
@@ -0,0 +1,75 @@
+
+
+requires "solidity-syntax.k"
+requires "configuration.k"
+
+
+module EXPRESSION
+
+imports SOLIDITY-SYNTAX
+imports CONFIGURATION
+
+rule [Allocate-State-Variables]:
+    <k> #allocate(N:Int, CN:Id, #varInfo(X:Id,E:Value,T:TypeName, #storage)) => . ...</k>
+    <account>
+        <acctID> #account(N, CN) </acctID>
+        <context> CONTEXT:Map => CONTEXT[X <- #storedVar(!N:Int, T, #storage)] </context>
+        <Storage> STORAGE:Map => STORAGE[!N <- E] </Storage>
+        ...
+    </account>
+
+rule [Allocate-Local-Variables]:
+    <k> #allocate(C:Int, CN:Id, #varInfo(X:Id, E:Value, T:TypeName, #mem)) => . ...</k>
+    <env> ENV:Map => ENV[X <- #storedVar(!N:Int, T, #mem)] </env>
+    <currentAccount> #account(C, CN) </currentAccount>
+    <localMem> MEM:Map => MEM[!N <- E] </localMem>
+
+rule
+    <k> X:Id => #read(X) ...</k>
+
+rule
+    <k> X:Id = I:Value => #write(X,I) ...</k>
+
+rule [Read-Local-Variables]:
+    <k> #read(X:Id) => I ...</k>
+    <env> ... X |-> #storedVar(N:Int, _:TypeName, #mem) ...</env>
+    <localMem> ... N |-> I:Value ...</localMem>
+
+rule [Read-State-Variables]:
+    <k> #read(X:Id) => I ...</k>
+    <env> ... X |-> #storedVar(N:Int, _:TypeName, #storage) ...</env>
+    <currentAccount> #account(C, CN) </currentAccount>
+    <account>
+        <acctID> #account(C, CN) </acctID>
+        <Storage> ... N |-> I:Value ...</Storage>
+        ...
+    </account>
+
+rule [Write-Local-Variables]:
+    <k> #write(X:Id,I:Value) => #end_Exp ...</k>
+    <env> ... X |-> #storedVar(N:Int, _:TypeName, #mem) ...</env>
+    <localMem> ... ((N |-> _:Value) => (N |-> I)) ...</localMem>
```

Comment:
Based on my experience, sometimes the suggested one will fail. If we have something that works. Why do we have to switch to another? This is not something wrong or not standard.

Reply:
Do you have the error messages/minimal examples for such cases? Let's keep the current version if the suggested one is buggy.

Reply:
In that case, the rule cannot be applied and the program will got stuck there. But by rewriting it to the current version, it works. Sometime it happens, but sometimes not.

Reply:
I recall we discussed this problem in the past. When you say that sometimes the problem happens but sometimes not, do you mean that it happens completely at random (e.g. running the same program N times in succession, without any other change) or that it happens after changing something apparently unrelated in the syntax/rules?

Actually it would be great if you could provide a minimal example that triggers this problem. That usually makes it easier to investigate and, if it turns out to be a K bug, to report it to the developers. 

Regarding the rule itself, I agree it's not technically wrong to use the more verbose version `((N |-> _:Value) => (N |-> I))`, but I also think the suggested version is better `N |-> (_ => I)`. 

My gut feeling is that, similarly to the previous issue with `type = Map` in `<account>`, this problem is indicating there's something incorrect somewhere, so while it may not be a priority right now, it's still worth investigating at some point.

Reply:
I agree that the writing is not wrong and I think it is K tool problem. You can take a look at the version in master. If you replace it with the suggested one, it cannot work. But in IMP example, it seems to work.

Reply:
These two expressions should be equivalent without changing other parts of the rule.

Reply:
See https://github.com/kframework/solidity-semantics/issues/30